src/library/scala/collection/immutable/Vector.scala

```diff
@@ -30,11 +30,23 @@ object Vector extends StrictOptimizedSeqFactory[Vector] {
 
   def from[E](it: collection.IterableOnce[E]): Vector[E] =
     it match {
-      case as: ArraySeq[E] if as.length <= 32 && as.unsafeArray.isInstanceOf[Array[AnyRef]] =>
```

Comment:
we should be able to reuse the same array without copying if it's exactly `Array[AnyRef]]`

```scala
scala> new Array[String](0).getClass == classOf[Array[AnyRef]]
res2: Boolean = false

scala> new Array[AnyRef](0).getClass == classOf[Array[AnyRef]]
res3: Boolean = true
```

```scala
as.unsafeArray.getClass == classOf[Array[AnyRef]]
```

Reply:
I thought about this, but thought this is a pretty niche scenario, who goes around calling `Vector.apply[AnyRef](...)`? Al alternate approach would be to special case the underlying `display0.clone()` logic (which is hit by `updated`) to be `display0.copyToArray(new Array[AnyRef](32))`. This would allow us to keep reusing the unsafeArray. (and maybe even in the case of primitives?)

Reply:
anything that creates a vector from a generic context with varargs will be an `ArraySeq` wrapping an `Array[AnyRef]`

Reply:
Ping @joshlemer, wdyt?

Reply:
@lrytz made a new draft PR https://github.com/scala/scala/pull/8276